### PR TITLE
Fix for norm() return `nan` when computing for a tensor of zeros.

### LINKED ISSUE #52712: Refractor `torch.linalg.norm`
Once PR #51099, `torch.linalg.norm` should be refactored to accomplish the following:

* Use out variants of functions internally to avoid resize and copy. For autograd support, this probably requires conditionally calling out-variants if the out-variant of `torch.linalg.norm` was called, and call non-out otherwise
* Frobenius norm should resolve to `linalg_vector_norm` with ord=2

### LINKED ISSUE #65599: Inconsistent subgradients of matrix frobenius norm
## 🐛 Bug

I found that the subgradients of matrix frobenius norm (when the norm is 0) are inconsistent across `torch.linalg.norm` and `torch.linalg.matrix_norm`. Specifically, while the former (`torch.linalg.norm`) can give more desired subgradients of 0, the latter (`torch.linalg.matrix_norm`) will produce NaN subgradients. Meanwhile, it is also weird that the results are even inconsistent between `torch.linalg.norm(x)` (default `ord`, which should be frobenius norm for matrix as stated in the documentation) and `torch.linalg.norm(x, ord='fro')`.

## To Reproduce

```python
>>> import torch
>>> a = torch.zeros(3, 3, requires_grad=True)
>>> torch.linalg.matrix_norm(a).backward()
>>> print(a.grad)
tensor([[nan, nan, nan],
        [nan, nan, nan],
        [nan, nan, nan]])
>>> a.grad.zero_();
>>> torch.linalg.norm(a, 'fro').backward()
>>> print(a.grad)
tensor([[nan, nan, nan],
        [nan, nan, nan],
        [nan, nan, nan]])
>>> a.grad.zero_();
>>> torch.linalg.norm(a).backward()
>>> print(a.grad)
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]])
```

## Expected behavior

The subgradients should be consistent across different functions that are by definition doing the same thing. And it seems to me that subgradients being 0 in this case is more desirable. Subgradients being NaN simply crashed my training.

## Environment

 - PyTorch Version: 1.9.0+cu111
 - OS: Ubuntu 20.04.1 LTS (x86_64)
 - How you installed PyTorch: pip
 - Build command you used: N/A
 - Python version: 3.8.3
 - CUDA/cuDNN version: 11.1

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @jianyuh @mruberry @walterddr @IvanYashchuk @xwang233

### PR DESCRIPTION
**Pull request by v0dro**
Fixes https://github.com/pytorch/pytorch/issues/65599.

The backward of the norm() function would go wrong as a result of calculating the `sqrt` when the sum of squares is 0. This PR mitigates this issue by taking different branches for `_out` and non-out calls to the `norm` function and defaults to the `vector_norm` function when `ord=2` or `"fro"`.

Additionally, this PR also partly fixes https://github.com/pytorch/pytorch/issues/52712. The copies could not be completely [removed](https://github.com/pytorch/pytorch/pull/71372/files#diff-db2b91c51cffd1fba933933ac20de256f5f0ced095890a1a49a17d73037001a2R2488) since the `linalg_vector_norm` function expects `result.dtype` and `opt_dtype` to always carry the same value, and also handles complex inputs differently. Changing the `linalg_vector_norm` is a out of scope of this PR.

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -2301,9 +2301,12 @@ Tensor &frobenius_norm_out(const Tensor& self,
     maybe_wrap_dims(dim_, self.dim());
     TORCH_CHECK(dim_[0] != dim_[1], "Expected dims to be different, got ", dim, " instead");
     if (self.is_complex()){
-      result_ = at::sqrt(at::sum(at::real(self.conj() * self), dim_, keepdim));
+      // Dont call sqrt if sum() is zero since it will produce nan with autograd.
+      auto sum = at::sum(at::real(self.conj() * self), dim_, keepdim);
```
Comment (Lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2305]: @anjali411 if we add `at::real` for real tensors we could merge both branches of the computation.

**Review by lezcano — CHANGES_REQUESTED** [review]
I might be mistaken, but I don't think this fix is correct. Consider a batch of two matrices, one of which is the zero matrix and the other one is an arbitrary matrix. The current implementation would compute the square root of the whole batch and the zero matrix would get NaNs as gradients.

**Comment by v0dro** [comment]
Ah yes you have a point. Will send a fix soon.

**Comment by nikitaved** [comment]
You will probably need to select non-zero/zero matrices and have two paths for them. You can check my last SVD PR on how it could get done with advanced indexing without any reshaping to 3D + index_select.

**Comment by Lezcano** [comment]
this is even simpler, as it can be done with `masked_fill`.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (anjali411) [reply aten/src/ATen/native/LinearAlgebra.cpp]: yeah sounds good! I forget if there's an outstanding PR for it. If not, I can create one!

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -2480,10 +2489,9 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, const op
     }
   } else {
     // 'ord' is int or None
-    std::vector<int64_t> dim_ = opt_dim.has_value() ? opt_dim.value().vec() : make_dim_list(ndim);
     if (!opt_num_ord.has_value() || dim_.size() == 1) {
       Tensor result_ = at::linalg_vector_norm(
-          self, opt_num_ord.value_or(2), opt_dim, keepdim, opt_dtype);
+                                              self, opt_num_ord.value_or(2), opt_dim, keepdim, opt_dtype);
```
Comment (Lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2494]: Join them?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 3]
```diff
@@ -2464,10 +2464,19 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, const op
...
   int64_t ndim = self.dim();
+  std::vector<int64_t> dim_ = opt_dim.has_value() ? opt_dim.value().vec() : make_dim_list(ndim);
+
+  bool str_ord_non_matrix_norm = false;
+
   if (opt_str_ord.has_value()) {
-    // 'ord' is string
     auto str_ord = opt_str_ord.value();
     check_str_ord_valid(str_ord, opt_dim, ndim);
+    str_ord_non_matrix_norm = (str_ord != "fro") && (dim_.size() != 2 || dim_.size() != 1);
```
Comment (Lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2474]: What does this `bool` flag mean? It's very difficult to read. It might be better to negate the meaning of the variable and call it `vector_norm_or_frob = (str_ord == "fro") || (dim_.size() == 2) || (dim_.size() == 1);`. Then dispatch to `linalg_vector_norm` in this case. The next question would be, why do we need `frobenius_norm_out` at all? That's just the usual vector norm.

Could you investigate whether this function has any benefits over `at::linalg_vector_norm` at all?

In fact, while looking at this, I found https://github.com/pytorch/pytorch/issues/52712. I think with little effort you could fix that issue as well as it's very relevant here.

**Review by lezcano — COMMENTED** [review]
So this is going in the right direction, but I think it can be improved. See the comments below.

**Comment by v0dro** [comment]
@Lezcano can you review this?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 4]
```diff
@@ -2482,12 +2481,23 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, const op
...
-          self, opt_num_ord.value_or(2), opt_dim, keepdim, opt_dtype);
-      // TODO: Resize and copy should be avoided with
-      //       https://github.com/pytorch/pytorch/issues/52712
-      at::native::resize_output(result, result_.sizes());
-      result.copy_(result_);
+      // The linalg_vector_norm() function must ensure that the result is the same dtype as
+      // opt_dtype. This check is done so that the _out variant can be called and copying
+      // can be avoided in case the opt_dtype is not supplied. linalg_vector_norm() also
+      // handles complex tensors separately using abs() and hence we use isComplexType().
+      if (isComplexType(self.scalar_type()) ||
```
Comment (Lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2488]: Why doesn't it work for complex dtypes?
Could you file an issue noting all these problems with `linalg_vector_norm`?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 5]
```diff
@@ -2650,18 +2660,48 @@ Tensor& linalg_matrix_norm_out(
...
+    return at::linalg_vector_norm(self, opt_ord.value_or(2), opt_dim, keepdim, opt_dtype);
+  }
+
   auto options = TensorOptions().dtype(opt_dtype.has_value() ? opt_dtype.value() : toRealValueType(self.scalar_type())).device(self.device());
   Tensor result = at::empty({0}, options);
-  return at::native::linalg_norm_out(
-      self, opt_ord, opt_dim, keepdim, opt_dtype, result);
+
+ if (dim_.size() == 2) {
+    _linalg_norm_matrix_out(result, self, opt_ord.value(), dim_, keepdim, opt_dtype);
```
Comment (Lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2673]: Why does it call the out variant?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 6]
```diff
@@ -2650,18 +2660,48 @@ Tensor& linalg_matrix_norm_out(
...
+
+ return result;
 }
 
 // Frobenius and nuclear norms
 Tensor linalg_norm(const Tensor& self, c10::string_view ord, optional<IntArrayRef> opt_dim, bool keepdim, optional<ScalarType> opt_dtype) {
+  check_str_ord_valid(ord, opt_dim, self.dim());
+
+  if (ord == "fro") {
+    return at::linalg_vector_norm(self, 2, opt_dim.value_or(IntArrayRef({0, 1})),
```
Comment (Lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2687]: Shouldn't this be `-2, -1`?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 7]
```diff
@@ -2650,18 +2660,48 @@ Tensor& linalg_matrix_norm_out(
...
   auto options = TensorOptions().dtype(opt_dtype.has_value() ? opt_dtype.value() : toRealValueType(self.scalar_type())).device(self.device());
   Tensor result = at::empty({0}, options);
-  return at::native::linalg_norm_out(
-      self, ord, opt_dim, keepdim, opt_dtype, result);
+  Tensor self_ = opt_dtype.has_value() ? self.to(opt_dtype.value()) : self;
+
+  // These _out functions need to be called since some of the tests expect the result to be of a particular dtype,
+  // which the non-out functions don't support.
+  if (opt_dim.has_value()) {
+    at::nuclear_norm_out(result, self_, opt_dim.value(), keepdim);
```
Comment (Lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2698]: This works almost by pure chance. This `_out` variant happens to be a composite function (_out variants should not be composite functions nor differentiable).
Could you leave a comment explaining this?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 8]
```diff
@@ -2650,18 +2660,48 @@ Tensor& linalg_matrix_norm_out(
...
+  } else {
+    TORCH_CHECK(false, "'dim' must specify 1 or 2 dimensions when order is numerical and input is "
+                "not 1-D or 2-D");
+  }
+
+ return result;
 }
 
 // Frobenius and nuclear norms
 Tensor linalg_norm(const Tensor& self, c10::string_view ord, optional<IntArrayRef> opt_dim, bool keepdim, optional<ScalarType> opt_dtype) {
```
Comment (Lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2683]: Also`, can't ord be `nuc` here? Where is this case treated?

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 5 above)
Comment (Lezcano) [inline aten/src/ATen/native/LinearAlgebra.cpp:2673]: why not `at::linalg_matrix_norm`?

**Review by lezcano — COMMENTED** [review]
There are still quite a few parts of the code that feel quite weird. Left a few comments.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 5 above)
Comment (v0dro) [reply aten/src/ATen/native/LinearAlgebra.cpp:2673]: This function is what implements the matrix norm. It is an internal function (not ATEN) and is just a 'function call' that does not get registered into autograd.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 4 above)
Comment (v0dro) [reply aten/src/ATen/native/LinearAlgebra.cpp:2488]: The `linalg_vector_norm` calls `abs()` on complex tensors for making sure that autograd produces the same result after the call to `abs()` in autograd. To do this, it will cast the `self` to the `in_dtype` and call `abs()`. If you pass the `result` tensor from the `out=` parameter into the `linalg_vector_norm_out` where the result is not already a complex tensor, the function will throw an error.

A potential fix is probably to cast the `result` tensor passed from `out=` to whatever type is necessary. However I am not sure if this is coherent with the rest of pytorch. 

See this:  https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L2559

**Comment by v0dro** [comment]
@mruberry can you review this?

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 4 above)
Comment (mruberry) [reply aten/src/ATen/native/LinearAlgebra.cpp:2488]: If I understand your analysis correctly I think you are correct, @v0dro. See https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch. The `out=` tensor must (conceptually) be copied to, so its dtype cannot be modified.

File: test/test_linalg.py [hunk 9]
```diff
@@ -1580,7 +1602,8 @@ def run_error_test_case(input, ord, dim, keepdim, error_type, error_regex):
             ((S, S), ['nuc'], 0, RuntimeError, r'order "nuc" can only be used if either len\(dim\) == 2'),
             ((S, S), ['fro'], 0, RuntimeError, r'order "fro" can only be used if either len\(dim\) == 2'),
             ((S, S), ['nuc'], (0, 0), RuntimeError, r'duplicate or invalid dimensions'),
-            ((S, S), ['fro', 0], (0, 0), RuntimeError, r'Expected dims to be different'),
+            ((S, S), ['fro'], (0, 0), RuntimeError, r'dim 0 appears multiple times'),
```
Comment (mruberry) [inline test/test_linalg.py:1605]: TODO (future PR): convert these error tests cases into proper ErrorInputs

File: torch/testing/_internal/common_methods_invocations.py [hunk 10]
```diff
@@ -9943,8 +9943,6 @@ def ref_pairwise_distance(input1, input2):
            sample_inputs_func=sample_inputs_linalg_norm,
            aten_name='linalg_norm',
            skips=(
-               # Pre-existing condition; Needs to be fixed
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:9946]: Nice!

**Review by mruberry — COMMENTED** [review]
Overall looks good to me but I'll let @Lezcano review the updated implementation more thoroughly

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 4 above)
Comment (Lezcano) [reply aten/src/ATen/native/LinearAlgebra.cpp:2488]: I think that there's two different issues here. So, on the one hand, we have that this function (for complex inputs) and `linalg_vector_norm` are functions from C to R (complex to real). As such, the out for **both** of them (in particular the `Tensor result_` returned by `at::linalg_vector_norm` two lines below) should be complex tensors. Hence, I don't understand why do we need the copy.

Now, the other thing to note here is that the 2-norm may be implemented as a complex pointwise operation (`mul`) + a complex reduction (`sum`) + a map to the reals (`abs`). There, it is true that we need an auxiliary tensor to hold the result of the first two operations. This auxiliary tensor should be complex. But then, I would assume that the implementation (if it's not a composite function) would use `abs_out` to copy the final result into the real tensor, and then it would return the `result` tensor. Now, note that this copy into a complex tensor should happen inside of `linalg_vector_norm_out`, so I still don't understand why this copy is necessary here.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 6 above)
Comment (Lezcano) [reply aten/src/ATen/native/LinearAlgebra.cpp]: @v0dro is this resolved?

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 7 above)
Comment (Lezcano) [reply aten/src/ATen/native/LinearAlgebra.cpp:2699]: The comment I was expecting went more along the lines of:
> // Even though these look like `_out` variants, they are in fact composite operations. While this is odd, it's fine to call them here, as they make `linalg_norm` still a composite operation.

**Review by lezcano — COMMENTED** [review]
A couple small comments, but this is looking great! Thanks @v0dro!
